### PR TITLE
Merge wheel trace

### PR DESCRIPTION
**Pull request by jnskkmhr**
Right now, only ROS1 support. 
Soon, I will add ROS2 version as well.

### CODE REVIEW

File: .gitmodules
Comment (AntoineRichard) [inline .gitmodules:1]: edit to match merge

File: WorldBuilders
Comment (AntoineRichard) [inline WorldBuilders:1]: WorldBuilder should be back to main.

File: cfg/environment/lunaryard_20m4SDG.yaml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 name: Lunaryard
-seed: 42  
+seed: 42
+init_with_robot: False
```
Comment (AntoineRichard) [inline cfg/environment/lunaryard_20m4SDG.yaml]: Not sure this is added line is a good idea as it locks the users into using the predefined robot. Maybe we should consider adding the option to select the desired robot? Another option could be to provide the robot path as an argument. If so, this should be done like that:

```yaml
robot:
  initialize_with_robot: False
  robot_path: some_path
```

File: cfg/environment/lunaryard_20m.yaml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 name: Lunaryard
-seed: 42  
+seed: 42
+init_with_robot: False
```
Comment (AntoineRichard) [inline cfg/environment/lunaryard_20m.yaml]: Not sure this is added line is a good idea as it locks the users into using the predefined robot. Maybe we should consider adding the option to select the desired robot? Another option could be to provide the robot path as an argument. If so, this should be done like that:

```yaml
robot:
  initialize_with_robot: False
  robot_path: some_path
```

File: cfg/environment/lunar_southpole.yaml
Comment (AntoineRichard) [inline cfg/environment/lunar_southpole.yaml:1]: I think this shouldn't be done this way. There should be a generic class that can handle any DEM and run it as a Geoclipmap. I really appreciate the work, but it's a bit rushed. So Instead of having a LSP lab there should be a GeoClipMap one. Static or dynamic.

File: cfg/environment/lunaryard_40m.yaml [hunk 3]
```diff
@@ -1,5 +1,6 @@
 name: Lunaryard
-seed: 42  
+seed: 42
+init_with_robot: False
```
Comment (AntoineRichard) [inline cfg/environment/lunaryard_40m.yaml]: Not sure this is added line is a good idea as it locks the users into using the predefined robot. Maybe we should consider adding the option to select the desired robot? Another option could be to provide the robot path as an argument. If so, this should be done like that:

```yaml
robot:
  initialize_with_robot: False
  robot_path: some_path
```

File: cfg/environment/lunaryard_80m.yaml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 name: Lunaryard
-seed: 42  
+seed: 42
+init_with_robot: False
```
Comment (AntoineRichard) [inline cfg/environment/lunaryard_80m.yaml]: Not sure this is added line is a good idea as it locks the users into using the predefined robot. Maybe we should consider adding the option to select the desired robot? Another option could be to provide the robot path as an argument. If so, this should be done like that:

```yaml
robot:
  initialize_with_robot: False
  robot_path: some_path
```

File: geo_clipmaps_demo_static_boulder.py
Comment (AntoineRichard) [inline geo_clipmaps_demo_static_boulder.py:1]: I am a bit confused as to why this isn't using the regular "run.py" pipeline. Is it just temporary?

File: src/environments/lunalab_deformable.py
Comment (AntoineRichard) [inline src/environments/lunalab_deformable.py:1]: I don't know if a separate controller is needed. Could these methods not be added into the regular controller?
It's up to discussion here.

File: src/environments/lunaryard_deformable.py
Comment (AntoineRichard) [inline src/environments/lunaryard_deformable.py:1]: As for the other file, not sure if a separate class should be used.

File: src/environments_wrappers/ros1/lunalab_deformable_ros1.py
Comment (AntoineRichard) [inline src/environments_wrappers/ros1/lunalab_deformable_ros1.py:1]: This will also link to the other comments about the deformable things. But I think there are a couple things that should be done differently on top of that. See comments in the file.

File: src/environments_wrappers/ros1/lunalab_deformable_ros1.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        )
+        self.robot_subs.append(
+            rospy.Subscriber(
+                "/Lunalab/Robots/Reset", String, self.resetRobot, queue_size=1
+            )
+        )
+        self.robot_subs.append(
+            rospy.Subscriber(
+                "/Lunalab/Robots/ResetAll", String, self.resetRobots, queue_size=1
+            )
```
Comment (AntoineRichard) [inline src/environments_wrappers/ros1/lunalab_deformable_ros1.py]: Is there a reason not to use the initialization from the mother class?

File: src/environments_wrappers/ros1/lunalab_deformable_ros1.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        p = [1.0, 1.0, 0.5]
+        q = [1, 0, 0, 0]
+        self.RM.addRobot(usd_path, robot_name, p, q, self.domain_id)
+    
+    def set_scene_view(self):
+        """
+        Sets the robot prim and rigid body view."""
+        robot_name = "ex1"
+        self.robot_prim.initialize(robot_name)
+        self.robot_prim_view.initialize(robot_name, self.scene)
```
Comment (AntoineRichard) [inline src/environments_wrappers/ros1/lunalab_deformable_ros1.py]: This has hardcoded robot name, it shouldn't. Also, is there a reason why this a ROS driven?

File: src/environments_wrappers/ros1/lunalab_deformable_ros1.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        contact_forces = self.robot_prim_view.get_net_contact_forces()
+        self.LC.deformTerrain(world_pose, contact_forces)
+    
+    def applyTerramechanics(self)->None:
+        """
+        Applies the terramechanics force and torque."""
+        velocities, omega = self.robot_prim.get_velocities()
+        sinkages = self.robot_prim.get_sinkages()
+        force, torque = self.TS.compute_force_and_torque(velocities, omega, sinkages)
+        self.robot_prim_view.apply_force_torque(force, torque)
```
Comment (AntoineRichard) [inline src/environments_wrappers/ros1/lunalab_deformable_ros1.py]: If this is non-ROS should it be there, or should it be offloaded to the controller class?

I can see that `applyTerramechanics` calls on the view, but you could also have an empty view and just add stuff to it.

File: src/environments_wrappers/ros1/lunaryard_deformable_ros1.py
Comment (AntoineRichard) [inline src/environments_wrappers/ros1/lunaryard_deformable_ros1.py:1]: This will share comments with the previous one. Actually now, that I'm looking at it, I feel like there could be only 1 ROS wrapper._

File: src/environments_wrappers/sdg/lunalab_deformable_sdg.py
Comment (AntoineRichard) [inline src/environments_wrappers/sdg/lunalab_deformable_sdg.py:1]: This trails back to the "do we need a unique file for deformable things, or can we have shared files?" question.
To be discussed.

File: src/environments_wrappers/sdg/lunaryard_deformable_sdg.py
Comment (AntoineRichard) [inline src/environments_wrappers/sdg/lunaryard_deformable_sdg.py:1]: Same

File: src/physics/terra_param.py
Comment (AntoineRichard) [inline src/physics/terra_param.py:1]: This should be in confs maybe?

File: src/robots/robot_view.py
Comment (AntoineRichard) [inline src/robots/robot_view.py:1]: I would give this file a different name since Isaac already has a robot class and a view for these things.
Though I understand why you chose that name. Maybe we could rework the robot thing a bit...

File: src/terrain_management/terrain_generation.py [hunk 8]
```diff
@@ -1,27 +1,31 @@
-__author__ = "Antoine Richard"
+__author__ = "Antoine Richard, Junnosuke Kamohara"
 __copyright__ = (
     "Copyright 2023, Space Robotics Lab, SnT, University of Luxembourg, SpaceR"
 )
 __license__ = "GPL"
 __version__ = "1.0.0"
 __maintainer__ = "Antoine Richard"
 __email__ = "antoine.richard@uni.lu"
-__status__ = "development"
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_generation.py:9]: Typo?

File: src/terrain_management/terrain_manager.py [hunk 9]
```diff
@@ -258,19 +299,38 @@ def update(self) -> None:
 
         pxr_utils.deletePrim(self._stage, self._mesh_path)
         self._sim_verts[:, -1] = np.flip(self._DEM, 0).flatten()
-        self.renderMesh(self._sim_verts, self._indices, self._sim_uvs)
+        with wp.ScopedTimer("mesh update"):
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_manager.py]: This should not stay in main once the rest of things are worked out.

File: src/terrain_management/terrain_manager.py [hunk 10]
```diff
@@ -242,6 +248,41 @@ def renderMesh(
...
+            mesh.GetFaceVertexCountsAttr().Set([3] * len(idxs))
+            UsdGeom.Primvar(mesh.GetDisplayColorAttr()).SetInterpolation("vertex")
+            pv = UsdGeom.PrimvarsAPI(mesh.GetPrim()).CreatePrimvar(
+                "st", Sdf.ValueTypeNames.Float2Array
+            )
+            pv.Set(uvs)
+            pv.SetInterpolation("faceVarying")
+
+        if colors:
+            mesh.GetDisplayColorAttr().Set(colors)
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_manager.py]: This is extremely similar to rendermesh, maybe it should be modified a tiny bit instead of having a whole new function that does the same thing? We could add a argument called `update: bool = False` that by-passes the last two lines of `renderMesh`.

**Review by AntoineRichard — CHANGES_REQUESTED** [review]
Overall, we need to take a bit more time to merge the Geoclipmap stuff. The software side of things needs to be thought through such that it can scale afterward.
Similarly regarding the terramechanics, we may want to do things somewhat differently.
In particular regarding the way methods are spread accross classes, and if we actually need new classes to implement these changes. Let's have a talk on Monday to sort this out. There are a couple other things, see comments per file.

File: cfg/environment/lunaryard_20m4SDG.yaml (hunk 1 above)
Comment (jnskkmhr) [reply cfg/environment/lunaryard_20m4SDG.yaml]: Okay, that sounds much more reasonable. 
I was going to add it later (first of all, hardcoding to make it work)

File: src/environments_wrappers/ros1/lunalab_deformable_ros1.py (hunk 5 above)
Comment (jnskkmhr) [reply src/environments_wrappers/ros1/lunalab_deformable_ros1.py]: I should inherit from the parent class. 
Thanks for pointing it out.

File: cfg/environment/lunaryard_20m.yaml (hunk 2 above)
Comment (jnskkmhr) [reply cfg/environment/lunaryard_20m.yaml]:
```yaml
robots:
-robot1:
  name:name
  path:name
  pose:pose
-robot2:
  name:name
  path:name
  pose:pose
```

**Comment by jnskkmhr** [comment]
@AntoineRichard 
Reworked robot code. 
Also merged some of redundant methods.

File: WorldBuilders
Comment (AntoineRichard) [inline WorldBuilders:1]: Do you have a matching PR on Worldbuilders?

File: cfg/environment/lunalab4SDG.yaml [hunk 11]
```diff
@@ -109,6 +109,28 @@ terrain_manager:
       min_elevation: -0.025
       z_scale: 1.0
       seed: ${....seed}
+    
+    deformation_engine:
+      enable: False
+      deformFrequencyInv: 10
```
Comment (AntoineRichard) [inline cfg/environment/lunalab4SDG.yaml]: Everything is using lower case and underscores, please match the style.

File: cfg/environment/lunaryard_20m.yaml [hunk 12]
```diff
@@ -233,9 +234,31 @@ terrain_manager:
...
+        distribution: sinusoidal
+        wave_frequency: 2.0
+      boundary_distribution: 
+        distribution: trapezoidal
+        angle_of_repose: 1.047
 
-    is_yard: False
-    is_lab: True
+    is_yard: True
+    is_lab: False
```
Comment (AntoineRichard) [inline cfg/environment/lunaryard_20m.yaml:260]: Not something to modify now, but we should change these to a string so that it simpler to configure.

File: cfg/environment/lunaryard_20m4SDG.yaml (hunk 1 above)
Comment (AntoineRichard) [reply cfg/environment/lunaryard_20m4SDG.yaml]: Still hardcoded?

File: src/robots/robot_parameter.py
Comment (AntoineRichard) [inline src/robots/robot_parameter.py:1]: This should be moved under the cfg folder.

File: src/terrain_management/terrain_manager.py [hunk 13]
```diff
@@ -215,7 +217,6 @@ def renderMesh(
             update_topology (bool): whether to update the mesh topology."""
 
         mesh = UsdGeom.Mesh.Get(self._stage, self._mesh_path)
-        self._mesh_path = self._og_mesh_path + "_" + str(self._id)
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_manager.py:218]: You need to be careful here, I don't know if they fixed it. But in 2022.2.1 their was an issue with computing colliders of mesh whose geometry changed. In short, the collider is cached and if the name is the same, then they do not recompute it. Please check that colliders are now updated properly before changing this line. I did raise an issue to them and they said it would be fixed soon, but I don't know if it ever has been.

File: src/terrain_management/terrain_manager.py [hunk 14]
```diff
@@ -240,37 +241,53 @@ def renderMesh(
         if colors:
             mesh.GetDisplayColorAttr().Set(colors)
 
-        self._id += 1
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_manager.py:243]: Same thing here.

**Review by AntoineRichard — CHANGES_REQUESTED** [review]

File: src/terrain_management/terrain_manager.py (hunk 13 above)
Comment (jnskkmhr) [reply src/terrain_management/terrain_manager.py:218]: Will check it. 
Thanks.

File: src/robots/robot_parameter.py
Comment (jnskkmhr) [reply src/robots/robot_parameter.py:1]: Not "config" folder?

Comment (jnskkmhr) [reply src/robots/robot_parameter.py:1]: I forgot to remove some of the parameters in cfg yaml. 
I will remove them.

Comment (AntoineRichard) [reply src/robots/robot_parameter.py:1]: Yes, it's what I meant, my bad :)

**Comment by jnskkmhr** [comment]
Hey @AntoineRichard 
Can you please review my PR?
I am thinking about having a separate file for deformation engine. 
This way, I could play with plotting etc w/o isaac.

File: src/terrain_management/terrain_manager.py (hunk 13 above)
Comment (jnskkmhr) [reply src/terrain_management/terrain_manager.py:218]: Yes you are right and fixed this part. 
2023.1.1 still has the issue about caching collider.

File: cfg/environment/lunalab4SDG.yaml (hunk 11 above)
Comment (jnskkmhr) [reply cfg/environment/lunalab4SDG.yaml]: updated.

File: WorldBuilders
Comment (jnskkmhr) [reply WorldBuilders:1]: Updated submodule. 
Just removed unnecessary code I wrote in clipper.py

File: src/terrain_management/terrain_generation.py [hunk 15]
```diff
@@ -494,18 +682,39 @@ def __init__(
...
+        self._dem_delta = None
+        self._multipass = None
+    
     def randomize(self) -> np.ndarray:
-        """
-        Generates a random terrain DEM with craters.
-
-        Returns:
-            np.ndarray: random terrain DEM with craters"""
-
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_generation.py]: Why are these lines removed?

File: src/terrain_management/terrain_generation.py [hunk 16]
```diff
@@ -463,6 +466,191 @@ def generateRandomTerrain(
...
+            heading = R.from_matrix(body_transform[:3, :3]).as_euler("zyx")[0]
+            projection_point = np.array(
+                [self.profile[:, 0]*np.cos(heading) - self.profile[:, 1]*np.sin(heading), 
+                self.profile[:, 0]*np.sin(heading) + self.profile[:, 1]*np.cos(heading)]).T + body_transform[:2, 3]
+            projection_points.append(projection_point)
+        return np.concatenate(projection_points)
+    
+    @staticmethod
+    def linear_func(force:np.ndarray, slope:float, intercept:float)-> np.ndarray:
+        return slope * force + intercept
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_generation.py]: missing function header. Is intercept intersect?

File: src/terrain_management/terrain_generation.py [hunk 17]
```diff
@@ -463,6 +466,191 @@ def generateRandomTerrain(
...
+            raise ValueError("Unknown model")
+    
+    def deform(self, DEM:np.ndarray, multipass:np.ndarray, body_transforms:np.ndarray, normal_forces:np.ndarray)-> np.ndarray:
+        """
+        Args:
+            DEM (np.ndarray): DEM to deform
+            body_transforms (np.ndarray): projected coordinates of rover's wheels (n, 4, 4)
+            normal_forces (np.ndarray): (dynamic) contact forces applied on rover's wheels (n, 3)
+            n is the number of wheels.
+        """
```
Comment (AntoineRichard) [inline src/terrain_management/terrain_generation.py]: Function header is incomplete. Is multipath a counter to store how many times a given pixel has been modified? If so maybe it should be named differently? Like num_pass?

File: src/terrain_management/terrain_generation.py (hunk 15 above)
Comment (jnskkmhr) [reply src/terrain_management/terrain_generation.py]: not sure, maybe i deleted accidentally

File: src/terrain_management/terrain_generation.py (hunk 16 above)
Comment (jnskkmhr) [reply src/terrain_management/terrain_generation.py]: yes. I mean b of ax+b

File: src/terrain_management/terrain_generation.py (hunk 17 above)
Comment (jnskkmhr) [reply src/terrain_management/terrain_generation.py]: yes, exactly. 
yearh, num_pass sounds more reasonable. 
okay, I will do that and then merge.

**Comment by jnskkmhr** [comment]
All good, merging. 
I will refactor deformation engine class later.
